src/lib/shim/shadow_sem.h

```diff
@@ -0,0 +1,35 @@
+#ifndef SHIM_SHADOW_SEM_H
+#define SHIM_SHADOW_SEM_H
+
+// Implements the same API as sem_init, sem_destroy, etc. from libc.
+//
+// This is a shared implementation used both in binary_spinning_sem and in
+// preload_libraries.
+
+#include <sys/time.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct shadow_sem_t {
```

Comment:
Yeah, this was semi-cargo-culted from glibc. I agree that this level of encapsulation is probably overkill for our purposes. Leaving for now until we decide whether to convert to C++, in which case we *would* need this sort of type.

Reply:
Done-ish. Unfortunately it turns out that C11 _Atomic is incompatible with C++, so I still had to hide that from the C++ code.

Reply:
OK, so an alternative:

Both C and C++ define the `atomic_uint_fast32_t` typedef in <stdatomic.h> and <atomic>, respectively. Unfortunately, C doesn't define an ordinary `atomic_uint32_t` whereas C++ does. `atomic_uint_fast32_t` is guaranteed to be _at least_ 32 bytes, but could be more -_- I'm guessing in practice that `sizeof(atomic_uint_fast32_t) == 4` for every platform we care about.

So, we could use a single definition of the struct and use conditional compilation to grab the right header, and then use `static_assert()` to make sure that we're still getting the right size and alignment.

Tbh, I'm not sure this change is worth making, so I'll leave it up to you.

Reply:
I couldn't get this to compile. It's a nice idea, but I think in the end it doesn't save us much since we still need the static asserts to validate we get the expected size and alignment.